gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,450 @@
+	LWLockRelease(tf_shared_state->bloom_set.lock);
+}
+
+static void
+dispatch_register_to_segments(List *dbids, List *uninitialized_segments)
+{
+	ListCell   *cell;
+	tracked_db_t *trackedDb;
+	CdbPgResults cdb_pgresults = {NULL, 0};
+
+	if (uninitialized_segments == NIL)
+		return;
+
+	foreach(cell, dbids)
+	{
+		trackedDb = (tracked_db_t *) lfirst(cell);
+
+		char	   *cmd = psprintf("select arenadata_toolkit.tracking_register_db(%u)", trackedDb->dbid);
+
+		CdbDispatchCommandToSegments(cmd,
+									 0,
+									 uninitialized_segments,
+									 &cdb_pgresults);
+
+		if (trackedDb->get_full_snapshot_on_recovery)
+		{
+			cmd = psprintf("select arenadata_toolkit.tracking_trigger_initial_snapshot(%u)", trackedDb->dbid);
+
+			CdbDispatchCommandToSegments(cmd,
+										 0,
+										 uninitialized_segments,
+										 &cdb_pgresults);
+		}
+	}
+}
+
+static void
+dispatch_register(bool dispatch_to_master, List *uninitialized_segments)
+{
+	Relation	rel;
+	SysScanDesc scan;
+	HeapTuple	tup;
+	List	   *dbids = NIL;
+	tracked_db_t *trackedDb;
+
+	rel = heap_open(DbRoleSettingRelationId, RowExclusiveLock);
+	scan = systable_beginscan(rel, InvalidOid, false, NULL, 0, NULL);
+	while (HeapTupleIsValid(tup = systable_getnext(scan)))
+	{
+		bool		isnull;
+		Datum		str_datum;
+		Datum		oid_datum;
+		ArrayType  *a;
+
+		str_datum = heap_getattr(tup, Anum_pg_db_role_setting_setconfig,
+								 RelationGetDescr(rel), &isnull);
+		if (isnull)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setrole,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) != InvalidOid)
+			continue;
+
+		oid_datum = heap_getattr(tup, Anum_pg_db_role_setting_setdatabase,
+								 RelationGetDescr(rel), &isnull);
+		if (DatumGetObjectId(oid_datum) == InvalidOid)
+			continue;
+
+		a = DatumGetArrayTypeP(str_datum);
+
+		if (is_db_tracked(a))
+		{
+			trackedDb = (tracked_db_t *) palloc0(sizeof(tracked_db_t));
+
+			trackedDb->dbid = DatumGetObjectId(oid_datum);
+			trackedDb->get_full_snapshot_on_recovery = full_snapshot_on_recovery(a);
+			dbids = lappend(dbids, trackedDb);
+		}
+	}
+
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	if (dbids != NIL)
+	{
+		ListCell   *cell;
+
+		if (dispatch_to_master)
+			dispatch_register_to_master(dbids);
+
+		dispatch_register_to_segments(dbids, uninitialized_segments);
+
+		foreach(cell, dbids)
+		{
+			pfree(lfirst(cell));
+		}
+
+		list_free(dbids);
+	}
+
+	LWLockAcquire(tf_shared_state->bloom_set.lock, LW_EXCLUSIVE);
+	tf_shared_state->bgworker_ready = true;
+	LWLockRelease(tf_shared_state->bloom_set.lock);
+}
+
+/* scan pg_db_role_setting, find all databases, bind blooms if necessary */
+static void
+arenadata_toolkit_worker(Datum main_arg)
+{
+	elog(LOG, "[arenadata toolkit] Starting background worker");
+
+	bool		master_initialized = false;
+
+	pqsignal(SIGHUP, tracking_sighup);
+	pqsignal(SIGTERM, tracking_sigterm);
+	pqsignal(SIGUSR1, tracking_sigusr1);
```

Comment:
Do we need handler for the SIGUSR1? It looks that its logic does nothing valuable. Can we remove it?